Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm

```diff
@@ -649,18 +664,32 @@ static void dataProviderReleaseInfoCallback(void* info)
         return 0;
     }
 
-    WTF::Semaphore dataSemaphore { 0 };
+    // Do not dispatch main runloop (again) and wait anymore if plugin has been destroyed.
+    if (hasBeenDestroyed())
+        return 0;
+
+    BinarySemaphore dataSemaphore;
     size_t bytesProvided = 0;
 
     RunLoop::main().dispatch([protectedLoader = Ref { *this }, position, count, buffer, &dataSemaphore, &bytesProvided] {
-        protectedLoader->getResourceBytesAtPosition(count, position, [count, buffer, &dataSemaphore, &bytesProvided](const uint8_t* bytes, size_t bytesCount) {
-            RELEASE_ASSERT(bytesCount <= count);
-            memcpy(buffer, bytes, bytesCount);
-            bytesProvided = bytesCount;
-            dataSemaphore.signal();
-        });
+        // protection against race when plugin has been destroyed in last runloop while this runloop has queued up.
+        if (!protectedLoader->hasBeenDestroyed()) {
+            protectedLoader->getResourceBytesAtPosition(count, position, [count, buffer, &dataSemaphore, &bytesProvided](const uint8_t* bytes, size_t bytesCount) {
+                RELEASE_ASSERT(bytesCount <= count);
+                memcpy(buffer, bytes, bytesCount);
+                bytesProvided = bytesCount;
+                dataSemaphore.signal();
+            });
+        }
     });
 
+    // One time notification observer to signal PDF thread to not wait for task dispatched in main runloop when main runloop is destructing the plugin.
+    NSNotificationCenter * __weak center = [NSNotificationCenter defaultCenter];
+    id token = [center addObserverForName:[PFDThreadIsExitingNotification stringByAppendingString:@(Thread::current().uid()).stringValue] object:nil queue:nil usingBlock:[&dataSemaphore, &center, &token](NSNotification *) {
+        dataSemaphore.signal();
+        [center removeObserver:token];
+    }];
```

Comment:
We can't pass a C++ lambda directly to an argument expecting a block.

Reply:
Let's rewrite this portion as

```suggestion
    RetainPtr<NSObject> token = [center addObserverForName:threadIsExitingNotificationName() object:nil queue:nil usingBlock:makeBlockPtr([&dataSemaphore, &center, &token](NSNotification *) {
        dataSemaphore.signal();
        [center removeObserver:token];
    }).get()];
```

Reply:
🤔 i saw some other places implemented something similar, wondering if why is the difference? sorry i am still a starter for c++.
i.e.:
```
m_sheetWindowWillCloseObserver = [[NSNotificationCenter defaultCenter] addObserverForName:NSWindowWillCloseNotification object:m_sheetWindow.get() queue:nil usingBlock:[weakThis = ThreadSafeWeakPtr { *this }] (NSNotification *) {
```

Reply:
As far as I know, as long as it builds, it's fine. I think we only need a block in cases where we capture non-copyable stuff. @aprotyas What's the concern here since it builds?

Reply:
The proposal makes the code more complicated so it needs a good reason.

Reply:
Yeah my bad about this suggestion, I probably didn't think it through. I had written some test code like this a while back that caused a runtime crash because I was providing a plain lambda and not a BlockPtr

Reply:
But that might have been specific to my use case. Since it builds and runs fine for you, keep it as is

Reply:
Thanks! But you are right for making `token` as `RetainPtr<NSObject>`, as we need to retain it before destruction.